### PR TITLE
Navigate to failed step in end device wizard

### LINKED ISSUE #3291: Navigate to failed step in end device wizard
#### Summary

We should navigate the user to wizard step that failed.

#### Why do we need this?

We make api requests only when submitting the last form in the end device wizard. Since the whole device creation flow is quite complicated and frontend validation might be not up to date to the actual validation rules defined on the backend, any request can fail. Currently, we display errors without navigating the user to the actual failed step.

Consider a simple case when a user trying to create a new device (OTAA) with already existing ID. With our current implementation the error will be presented at the Join Settings step, while the actual step caused the error is the Basic Settings step.

#### What is already there? What do you see now?

1. End device wizard
2. API in the JS SDK

#### What is missing? What do you want to see?

1. Functionality in `<Wizard />` to navigate to arbitrary step.
2. Functionality to check the component that returned an error.

#### How do you propose to implement this?

TBA

#### How do you propose to test this?

e2e tests, manual testing

#### Can you do this yourself and submit a Pull Request?

yes

### PR DESCRIPTION
**Pull request by bafonins**
#### Summary

Closes https://github.com/TheThingsNetwork/lorawan-stack/issues/3291

wizard-error

<details>
<summary>Submit bar fix screenshots</summary>
Screenshot 2020-10-22 at 14 02 22
Screenshot 2020-10-26 at 16 11 23
Screenshot 2020-10-26 at 16 11 29

</details>

#### Changes

- Add step id and rely on it instead of step index
- Extend wizard actions
- Export stack configuration map from the js sdk
- Augment errors originating in the js sdk to include additional information about failed request (method, url, failed component, params)
- Change `net ID` to `NetID` in the join settings form (see https://github.com/TheThingsNetwork/lorawan-stack/pull/3065#issuecomment-705409095)
- Fix `<SubmitBar />` buttons for small screens

##### Regressions

Navigation in the end device wizard.

#### Notes for Reviewers

1. `request_details` in errors - Can also be helpful for https://github.com/TheThingsNetwork/lorawan-stack/issues/2681.
2. Current implementation works fine for various scenarios I tested, however it is not possible to test all possible inputs so there might be issues with navigating to the actual failed step because some values are accumulated from previous steps (e.g. device `ids`). Proper solution to avoid such uncertainty would be to submit forms on each step instead of doing this only after the last one.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: sdk/js/src/api/http.js [hunk 1]
```diff
@@ -112,7 +113,16 @@ class Http {
...
       if ('response' in err && err.response && 'data' in err.response) {
-        throw err.response.data
+        const error = cloneDeep(err.response.data)
+        error.request_details = {
+          url: get(err, 'response.config.url'),
+          method: get(err, 'response.config.method'),
+          params: get(err, 'response.config.params'),
+          base_url: get(err, 'response.config.baseURL'),
+          stack_component: parsedComponent,
+        }
```
Comment (kschiffer) [inline sdk/js/src/api/http.js]: Not too happy about this. I'd prefer to leave the error responses unmanipulated.

IMO, a better option would be to change the output for device CRUD functions in the SDK so that when an error occurs, they return an object of errors keyed by component ID. Currently, we just return the first error that occurred.

File: sdk/js/src/api/http.js [hunk 2]
```diff
@@ -141,7 +151,7 @@ class Http {
    * Extracts The Things Stack component abbreviation from the endpoint.
    *
    * @param {string} endpoint - The endpoint got for a request method.
-   * @returns {string} One of {is|as|gs|js|ns}.
+   * @returns {string} One of {is|as|gs|js|ns|edtc|qrg|gcs}.
```
Comment (kschiffer) [inline sdk/js/src/api/http.js]:
```suggestion
   * @returns {string} The stack component abbreviation.
```

We will also forget to update this next time for sure.

File: pkg/webui/components/wizard/wizard.js [hunk 3]
```diff
@@ -34,88 +36,120 @@ const reducer = (state, action) => {
...
-  const { currentStep, steps, snapshots } = state
-
-  const stepsCount = steps.length
+  const { currentStepId, steps, snapshots, error } = state
 
   const stepsInit = React.useCallback(steps => {
     dispatch({ type: INIT, steps })
   }, [])
+  const completeStep = React.useCallback(values => {
+    console.log('kekster mani suka')
```
Comment (kschiffer) [inline pkg/webui/components/wizard/wizard.js]: Remove.

**Review by kschiffer — CHANGES_REQUESTED** [review]
The changes regarding the next/prev button layouts in small screensizes don't seem to be committed here.

Please implement it like this:
- Retain the text of the "next step" and "Submit" button
- Hide the text of the "back" button
- Overwrite the default value of the `min-width` for the back button to `auto`

File: sdk/js/src/api/http.js (hunk 1 above)
Comment (bafonins) [reply sdk/js/src/api/http.js]: > Not too happy about this.

Me neither, but its the simplest solution that can be changed at any moment, does not require any api changes in the console/sdk and uses functionality that we already have in the sdk. Also, this provides additional debug info (http method, full url and target stack component can be helpful).

> IMO, a better option would be to change the output for device CRUD functions in the SDK so that when an error occurs, they return an object of errors keyed by component ID. 

Then I would expect device CRUD methods to return successful results in a similar shape.
```
{
  is: {... device fields stored in is ...},
  as: {... device fields stored in as ...},
  ...
}
```

We can adjust this by changing signatures of device crud methods, for example
```
async create(applicationId, device, mask = Marshaler.fieldMaskFromPatch(device))
// can be changed to:
async create(applicationId, device, mask = Marshaler.fieldMaskFromPatch(device), mergeResult = true)
// or
async create(applicationId, device, mask = Marshaler.fieldMaskFromPatch(device), { mergeResult = true })
// or
async create(applicationId, device, { mask = Marshaler.fieldMaskFromPatch(device), mergeResult = true })
```

Basically when `mergeResult` is `true` then both merged result and first error is returned, otherwise we return results split by component, see above. This will require some changes to end device redux primitives.

I can add utility functions to `lib/errors/utils.js` to get the target component and use those instead of directly accessing `request_details` in the device-add wizard. So later we can switch to other implementation (we can check for failed component by error message or anything else)

Comment (kschiffer) [reply sdk/js/src/api/http.js]: Indeed, I was thinking of a solution like that. Can you put this into a new issue and add a `TODO:` here?

Note that when manipulating the error object, these things will be part of the displayed error details in the Console. Meaning that they won't represent the backend error 1:1 anymore. This might confuse others when trying to debug.

**Review by kschiffer — APPROVED** [review]
See my comment about adding a `TODO:`
